File: src.JAVA/src/swa/ariadne/logic/RoundRobinFlooder.java
package swa.ariadne.logic;

import java.util.*;

import swa.ariadne.model.*;

/**
 * A {@link IMazeSolver MazeSolver} with many concurrent paths.
 * Visits all neighbor squares of the current path's end before advancing to the next path.
 */
public
class RoundRobinFlooder extends FlooderBase
{
    //--------------------- Constructor

    /**
     * Constructor.
     * @param maze The problem to be solved.
     * @param mazeDrawer An object that can draw the maze. 
     */
    public RoundRobinFlooder(Maze maze, IMazeDrawer mazeDrawer)
    {
        super(maze, mazeDrawer);
    }
    
    //--------------------- FlooderBase Implementation

    @Override
    protected int selectPathIdx()
    {
        // Always select the first open path.
        return 0;
    }

    @Override
    protected WallPosition selectDirection(MazeSquare sq, List<WallPosition> openWalls)
    {
        return openWalls.get(0);
    }

}

File: src.CS/SWA.Ariadne.Logic/RoundRobinFlooder.cs
using System;
using System.Collections.Generic;
using System.Text;
using SWA.Ariadne.Model;
using SWA.Ariadne.Model.Interfaces;

namespace SWA.Ariadne.Logic
{
    /// <summary>
    /// A MazeSolver with many concurrent paths.
    /// Visits all neighbor squares of the current path's end before advancing to the next path.
    /// </summary>
    internal class RoundRobinFlooder : FlooderBase
    {
        #region Constructor

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="maze"></param>
        /// <param name="mazeDrawer"></param>
        public RoundRobinFlooder(Maze maze, IMazeDrawer mazeDrawer)
            : base(maze, mazeDrawer)
        {
        }

        #endregion

        #region Runtime methods

        /// <summary>
        /// Select an index within the flooder's list of open paths.
        /// </summary>
        /// <returns></returns>
        protected override int SelectPathIdx()
        {
            int result = 0;

            while (result < list.Count && !IsSelectablePathIdx(result))
            {
                ++ result;
            }

            return result;
        }

        /// <summary>
        /// Select one of the open walls leading away from the given square.
        /// </summary>
        /// <param name="sq1"></param>
        /// <param name="openWalls"></param>
        /// <returns></returns>
        protected override WallPosition SelectDirection(MazeSquare sq1, List<WallPosition> openWalls)
        {
            return openWalls[0];
        }

        #endregion
    }
}
